base/compiler/tfuncs.jl

```diff
@@ -484,14 +484,27 @@ function typeof_tfunc(@nospecialize(t))
 end
 add_tfunc(typeof, 1, 1, typeof_tfunc, 0)
 
-function typeassert_tfunc(@nospecialize(v), @nospecialize(t))
-    t = instanceof_tfunc(t)[1]
-    t === Any && return v
+function typeassert_instance(@nospecialize(v), @nospecialize(t))
     if isa(v, Const)
         if !has_free_typevars(t) && !isa(v.val, t)
             return Bottom
         end
         return v
+    elseif isa(v, PartialStruct)
+        has_free_typevars(t) && return v
+        widev = widenconst(v)
+        newT = typeintersect(widev, t)
+        if newT === widev
+            return v
+        end
```

Comment:
```suggestion
        if widev <: t
            return v
        elseif typeintersect(widev, t) === Bottom
            return Bottom
        end
```

Reply:
Calling `typeintersect` here would force us to call `abstract_apply` on `tuple_tfunc` over the result, to recompute `widev` from `newT`. That's awkward (sometimes more accurate, sometimes less), so let's avoid that.

Reply:
I don't understand what you're saying here.